Add MSBuild properties for common ILLink arguments

sbomer opened the pull request:
For now this just adds private properties to make it easier for SDK components to set ILLink arguments (instead of using `_ExtraTrimmerArgs`). Most of them are not set by default, with the exception of `_TrimmerDefaultAction` which replaces the `-c` and `-u` flags.

I don't think there is much to test here, since most of the new properties are not tied to existing SDK logic. An existing test checks the behavior for `_TrimmerDefaultAction`, and I have added a sanity check that the boolean options for the optimizations are actually used by validating that they produce errors if they're not boolean. The functionality of the options themselves is tested in https://github.com/mono/linker.

fixes https://github.com/dotnet/sdk/issues/11006.

sbomer committed:
Add MSBuild properties for common ILLink arguments

dsplaisted commented:
Is there a PR for ILLink that you can link to where these options were added to the task?

sbomer commented:
@dsplaisted here it is: https://github.com/mono/linker/pull/1057

dsplaisted commented:
Does this complete #11006?

dsplaisted commented:
Looks like some tests are failing, probably because they target .NET Core 3.x and are referencing a package which now requires .NET 5.  The tests may need to be retargeted, or possibly split out so we are testing both .NET Core 3.x and .NET 5 scenarios.

sbomer commented:
Yes, this fixes https://github.com/dotnet/sdk/issues/11006. The test failures had me pretty confused for a bit - I think the issue is that the referenced project is shared across all of those tests, which wasn't the intention. When I created it for 5.0 in the new test, some of the existing testcases were picking that up (probably when running in parallel). I'm going to fix it so that there's a separate reference project for each testcase.

sbomer committed:
Place referenced project into testcase directory

sbomer committed:
Avoid creating a net5.0 packagereference

The test infrastructure assumes that a project with a given name will be
packaged for just one TFM, placing it in an output directory that doesn't
include the TFM.

sbomer committed:
Don't target net5.0 on full framework MSBuild

sbomer commented:
@dsplaisted PTAnotherL

dsplaisted commented on src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToRunILLink.cs:
```diff
@@ -441,8 +453,35 @@ private void EnableNonFrameworkTrimming(XDocument project)
+            string mainProjectName,
+            string referenceProjectName = null,
+            bool usePackageReference = true,
+            [CallerMemberName] string callingMethod = "")
```
If you are calling this method from a `Theory`, then the different instances will still collide on the path.  The pattern we generally use for this is to also pass in an optional `identifier` parameter, which can be a concatenation of the `Theory` parameters.

sbomer replied:
Thanks!

sbomer committed:
Prevent project location conflict for theories
